### ISSUE TITLE
Support changing folder icons [$15]

### ISSUE THREAD

**Issue by codygarver** [elementary/files#171]
Support changing folder icons.

Bounty: https://www.bountysource.com/issues/21220585-possibility-to-change-folder-icons-15

**Comment by aljelly** [comment]
In addition to (or instead of?) being able to choose folder icons for user folders, it would also be really cool if you could show it as a sort of "badge" on top of the default folder icon. Here's a mockup:

![mockup1](https://user-images.githubusercontent.com/10395308/37195710-e2aabcb0-236b-11e8-9b27-2d658958d22d.png)

![mockup1boundaries](https://user-images.githubusercontent.com/10395308/37195713-e4d95244-236b-11e8-8d7f-6f10a07677d4.png)

(The icon and border in the mockup is not perfectly square, it's about 1px less wide, it should be 32px x 32px)

Basically a custom icon positioned over the default folder icon. For the folder icon size shown in the mockup, 32px x 32px would be an appropriate size for a badge icon. The badge sizes for other folder icon sizes would probably need to be considered.

For the smallest sizes where the special folder icons (Documents, Music, Videos etc.) turn into symbolic icons (like a piece of paper, a musical note, film), you could show the badge icon as the icon for the folder.

(This is 200% a wishlist addition and shouldn't be made a priority at all.)

**Comment by Vento09** [comment]
It could be like webpin. When we were to open the properties of the folder, you could have the option to change the folder icon like this:

![propriedades-pantheon](https://user-images.githubusercontent.com/30022019/37406405-da78416c-2775-11e8-8a47-f52107b4a028.png)

(The image was not good but what matters is the intetion)

It could be possible with archives also, mainly with .appimages archives

**Comment by jeremypw** [comment]
@danrabbit Is there any scope for providing a limited list of (symbolic?) icons that a user can choose to embed onto the standard folder icon? e.g. those in `/usr/share/icons/elementary/apps`?
Or is this a complete non-starter?

**Comment by jeremypw** [comment]
Colored star overlays could be used as an alternative to color tags and could be stored in metadata rather than a separate database.

**Comment by danrabbit** [comment]
@jeremypw could do, I think the main thing here is figuring our first what the problem being solved is and if this is actually the best way to solve it. I think when this has come up before nobody has been able to articulate what this feature is meant to help with. It seems like a solution in search of a problem

For example, I'm not sure "Apps" makes sense since sandboxed apps will have limited file system access and if they should have folders outside of `.local` we probably would want to badge them automatically instead of having users do it manually. But until there's a clear understanding of what this feature is intended to be used for, it's not clear how it should be solved.

**Comment by jeremypw** [comment]
I presume the main use is to indicate what kind of files the folder contains, or which app or activity they relate to, where that is not already covered by the existing special user directories.  I agree that in the case of apps it would be better for them to be automatically badged. Need to bear in mind that emblem space is limited and the size of emblems is very small so overlaying the icon allows for more complex badges and easier identification.

Also, presuming that the existing color tags have a valid use (do they?), overlaying the icon would be a more identifiable alternative to that in the case of folders.

I guess it boils down to having a way of identifying a folder other than reading the filename, which may be easier for some people.

**Comment by jeremypw** [comment]
@aljelly  Do you have specific use cases in mind?

**Comment by JindrazPrahy** [comment]
>I think the main thing here is figuring our first what the problem being solved is and if this is actually the best way to solve it. I think when this has come up before nobody has been able to articulate what this feature is meant to help with. It seems like a solution in search of a problem

So the issue I had that led me to want folder icons (and to seek out this issue and comment on it) is the following:
1a) I open the file manager and want to quickly navigate to a desired folder, but I don't want to read the folder names, because they are too small (I don't have my glasses on at the moment), or it takes too much time to read them (basically, since the Folders are sorted alphabetically, I need to do a manual binary search to find the folder I want).
If the folders have their custom icons, I can instantly recognise the desired folder, because I see more things at once, but I can't read more things at once.

This exact argument has already been mentioned in the related PR:
>I guess it is to give an "at a glance" indication of the type folder contents without having to read the name (which might be in small text) 🤷

1b) I guess I can choose to highlight the name of the file in color, but it doesn't seem to me to be as clear as having folder icons (also, users may be colorblind).

Different reasons why this proposal might be a good idea:
2a) The folder icon provides me with a look at what is inside the folder. This can be literal, I can put a thumbnail of a file inside the folder as an icon.* Or conceptual, I can put an icon that represents the contents of the folder. Thus, to get the overview of the contents of the folder, I don't even need to open it, the icon gives me a conceptual hint.

2b) Let's say I open my Desktop folder with 20 subfolders. The UI is then filled with 20 identical images of folders and small text fields. Most of my field of vision is then filled with elements that do not provide much information value - that seems like wasted space.

2c) Having custom folder icons is just pretty, I want to be able to customise them. Ex.: Let's say there is a GitLab project. The GitLab project has an icon. When I clone it, wouldn't it be nice if I could set the icon to be the same as the GitLab project? Ex2.: Let's say I work in an organisation and I put the files related to work into a folder. The organisation has a logo, so why not change the folder icon to be the logo: seems pretty nice to me.

2d) There is already a bounty on this issue, so people probably want it?

*A related feature would be to automatically put thumbnails of files inside the folder into the folder icon. It's a different feature though.

**Comment by JindrazPrahy** [comment]
>a folder could have any icon 
Well this design language is already present in the app. In the bookmarks bar on the left side, the home icon is a house, the music icon is a note, the videos icon is a film strip, the Downloads icon is a little arrow... And it doesn't cause confusion (I think).

Also, since the user would set the icon themselves, I find it hard to believe that they would get confused by *their own* deliberate action (although this is just my conjecture and is not backed up by testing).

**Comment by jeremypw** [comment]
> In the bookmarks bar on the left side, the home icon is a house, the music icon is a note, the videos icon is a film strip, the Downloads icon is a little arrow.

Bookmark entries area assumed to be folders (although admittedly in Files this is not necessarily true). I have sympathy with the stance that users should be allowed to do what they want so long as it is not actively harmful though.

**Comment by jeremypw** [comment]
I not that Nautilus (latest) does support replacing the folder icon with an arbitrary custom image. Not sure how secure that is.

**Comment by zachari043** [comment]
Users want this option, no matter why they want it! For me I want my files to be beautiful and organized, no more no less it's pleasing me.  I must have a reason?  +1 on this for 6.4 !

**Comment by jeremypw** [comment]
This is unlikely to implemented in the Gtk3 version now. The Gtk4 version requires a very different way of displaying file items so I'll try to build in the possibility even though whether it used is still an open design decision.

**Comment by danirabbit** [comment]
Looks like GNOME is doing a survey regarding this feature so we might get some valuable data from there https://discourse.gnome.org/t/survey-custom-file-thumbnail-usage/18241
